clang/lib/CIR/Dialect/Transforms/TargetLowering/LowerFunction.cpp

```diff
@@ -246,6 +263,31 @@ Value castReturnValue(Value Src, Type Ty, LowerFunction &LF) {
     return LF.getRewriter().create<LoadOp>(Src.getLoc(), Cast);
   }
 
+  // Otherwise do coercion through memory. This is stupid, but simple.
+  if (auto addr = findAlloca(Src.getDefiningOp())) {
+    auto &rewriter = LF.getRewriter();
+    auto *ctxt = LF.LM.getMLIRContext();
+    auto ptrTy = PointerType::get(ctxt, Ty);
+    auto voidPtr = PointerType::get(ctxt, mlir::cir::VoidType::get(ctxt));
+
+    AllocaOp tmp;
+    {
+      mlir::PatternRewriter::InsertionGuard guard(rewriter);
```

Comment:
How come we need a guard just around the creation of the alloca but not anything else?

Reply:
Because we want to insert one alloca after another one, and don't insert it in some random place, e. g. in the middle of a block.  Thus, guard preserve current insertion point and restore it after the scope end.

Reply:
Well, I would glad to. But this function is from one of the CIR builders ( I don't recall exactly, either from CIRBaseBuilder or from CIRGenBuilder) and here we use rewriter, and mix both of them will make it harder to read, or.. what do you think?

Reply:
That makes sense, I'm just wondering how it works, given that you're not explicitly changing the insertion point in this block (at least as far as I can see). Does just creating the InsertionGuard set the insertion point to the right place automatically?

Reply:
This is a bit confusing and I might be wrong, but seems like there are two different inserters going on here (a) the rewriter and (b) the general builder inserter - I'm not sure they are kept in sync (I've seen cases where it isn't). I wouldn't be surprised if this is working out of luck.

In theory, after the scope ends, the insertion point will be back to before the `AllocaOp`, but then the bitcasts work because that isn't tracked via the rewriter, otherwise you'd be inserting it before the Alloca. What comes next does uses the rewriter again, but seems like it's being places after the alloca, hence my confusion. What am I missing?

Reply:
> mix both of them will make it harder to read, or.. what do you think?

very confusing, I've missed that during review, thanks for pointing out

Reply:
There is no general builder as far as I see,  everywhere we use  just the rewriter passed in the constructor of either of `LowerFunction` or `LowerModule` classes. 

But in the same time looks you're right - I didn't set the insertion point for `AllocaOp` properly. 

So now I just explicitly save and restore the point before and after the alloca insertion and then continue with the next operations.